src/main/java/edu/harvard/iq/dataverse/marketplace/controller/api/AuthController.java

```diff
@@ -0,0 +1,260 @@
+        
+        String jwt = jwtUtils.generateJwtToken(authentication);
+
+        UserDetailsImpl userDetails = (UserDetailsImpl) authentication.getPrincipal();
+        List<String> roles = userDetails.getAuthorities().stream().map(item -> item.getAuthority())
+                .collect(Collectors.toList());
+
+        return ResponseEntity
+                .ok(new JwtResponse(jwt,
+                        userDetails.getId(),
+                        userDetails.getUsername(),
+                        userDetails.getEmail(),
+                        roles));
+    }
+
+    @PostMapping("/signup")
+    @PreAuthorize("hasAuthority('ADMIN')")
+    @AuthAPIDocs.Signup
+    public ResponseEntity<?> registerUser(@Valid @RequestBody SignupRequest signUpRequest) {
+
+        ServerMessageResponse messageResponse = null;
+
+        if (userRepository.existsByUsername(signUpRequest.getUsername())) {
+            messageResponse = new ServerMessageResponse(HttpStatus.BAD_REQUEST,
+                    "Username is already taken.",
+                    "Error during the creation of the user, a user with this username already exist.");
+
+            return ResponseEntity.badRequest().body(messageResponse);
+        }
+
+        if (userRepository.existsByEmail(signUpRequest.getEmail())) {
+            messageResponse = new ServerMessageResponse(HttpStatus.BAD_REQUEST,
+                    "Email is already taken.",
+                    "Error during the creation of the user, a user with this email already exist.");
+            return ResponseEntity.badRequest().body(messageResponse);
+        }
+
+        User user = new User(signUpRequest.getUsername(),
+                signUpRequest.getEmail(),
+                encoder.encode(signUpRequest.getPassword()));
+
+        userRepository.save(user);
+
+        messageResponse = new ServerMessageResponse(HttpStatus.OK,
+                "User registered successfully.",
+                "The user was successfully registered.");
+
+        return ResponseEntity.ok(messageResponse);
+    }
+
+    @PostMapping("/role")
+    @PreAuthorize("hasAuthority('ADMIN')")
+    @AuthAPIDocs.RoleCreationRequest
+    public ResponseEntity<?> createRole(@Valid @RequestBody RoleCreationRequest roleCreationRequest) {
+        
+        if(roleRepository.existsByName(roleCreationRequest.getRoleName().toUpperCase())) {
+            return ResponseEntity.badRequest().body(new ServerMessageResponse(HttpStatus.BAD_REQUEST,
+                    "Role already exists.",
+                    "Error during the creation of the role, a role with this name already exist."));
+        }
+        
+        Role role = new Role();
+        role.setName(roleCreationRequest.getRoleName().toUpperCase());
+        roleRepository.save(role);
+        return ResponseEntity.ok(new RoleCreationResponse(role));
+    }
+
+    @PreAuthorize("hasAuthority('ADMIN')")
+    @PostMapping("/role/{roleId}/user/{userId}")
```

Comment:
```suggestion
    @PostMapping("/role/{roleId}/user/{userId}")
    @PreAuthorize("hasAuthority('ADMIN')")
```

Maybe swap these and always put `@PostMapping` first, like above.

Or always put `@PreAuthorize` first? For consistency. 🤷

Reply:
I think I will put always the PreAuthorize first since it may or may not be, the post mapping will be always there in top of the docs for consistency.